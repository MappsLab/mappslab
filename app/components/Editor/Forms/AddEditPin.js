// @flow
import React from 'react'
import styled from 'styled-components'
import type { NewPinType } from 'Types'

const Label = styled.label`
	font-size: 18px;
	display: block;
`

const Input = styled.input`
	font-size: 16px;
	border: 1px solid black;
	padding: 3px 5px;
	margin: 10px 0;
`

const Textarea = styled.textarea`
	font-size: 16px;
	border: 1px solid black;
	padding: 3px 5px;
	margin: 10px 0;
`

const Button = styled.button`
	display: block;
	margin: 0 auto;
	border: 2px solid gray;
	border-radius: 2px;
	font-size: 12px;
	padding: 4px;
`

const FormContainer = styled.div`
	border: 1px solid gray;
	box-shadow: 0px 2px 2px rgba(0, 0, 0, 0.2);
	border-radius: 2px;
	position: absolute;
	width: 400px;
	bottom: 120px;
	left: calc(50% - 200px);
	background: white;
	z-index: 110;
	padding: 10px;
	display: flex;
	flex-direction: column;
	justify-content: center;
	align-items: center;
`

/**
 * AddEditPin
 */

type Props = {
	handleSubmit: ({}) => void,
	pin: NewPinType,
}

const AddEditPin = ({ handleSubmit, pin }: Props) => {
	const { title, description } = pin
	return (
		<FormContainer>
			<form onSubmit={handleSubmit}>
				<Label>What's this pin called?</Label>
				<Input type="text" id="title" name="title" value={title} />
				<Textarea id="description" name="description" value={description || ''} />
				<Button type="submit">submit</Button>
			</form>
		</FormContainer>
	)
}

export default AddEditPin
