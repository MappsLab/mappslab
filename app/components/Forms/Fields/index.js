import BasicField from './BasicField'
import ImageField from './ImageField'
import Textarea from './Textarea'

module.exports = {
	BasicField,
	ImageField,
	Textarea,
}
