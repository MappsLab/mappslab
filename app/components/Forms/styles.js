// @flow
import { css } from 'styled-components'

/* Forms */

export const formWrapperStyles = css`
	display: flex;
	flex-direction: column;
	justify-content: center;
	align-items: center;
	text-align: center;
	${({ disabled }) =>
		disabled &&
		`
		opacity: 0.5;
		pointer-events: none;
	`};
`

export const formMessageStyles = css`
	font-size: ${({ theme }) => theme.text.h4};
`

export const formErrorStyles = css`
	${formMessageStyles};
	color: ${({ theme }) => theme.colors.red};
`

/* Fields */

export const fieldWrapperStyles = css`
	display: ${({ hidden }) => (hidden ? 'none' : 'flex')};
	flex-direction: column;
	align-items: flex-start;
	width: 100%;
`
// margin: ${({ theme }) => `${theme.layout.spacing.single} 0 ${theme.layout.spacing.double}`};

export const labelStyles = css`
	font-weight: ${({ theme }) => theme.text.semi};
	font-size: ${({ theme }) => theme.text.h4};
	text-transform: uppercase;
	margin-bottom: ${({ theme }) => theme.layout.spacing.half};
`

export const helpTextStyles = css`
	margin-top: ${({ theme }) => theme.layout.spacing.eighth};
	font-size: ${({ theme }) => theme.text.h4};
	color: ${({ theme }) => theme.colors.middleGray};
	text-align: left;
`

export const validationErrorStyles = css`
	${helpTextStyles};
	color: ${({ theme }) => theme.colors.red};
`

export const inputStyles = css`
	font-size: ${({ theme }) => theme.text.p};
	width: 100%;
	padding: ${({ theme }) => theme.layout.spacing.quarter};
	border-bottom: 2px solid;
	border-color: ${({ active, theme }) => (active ? '' : theme.colorslightGray)};
	transition: 0.3s;
`
