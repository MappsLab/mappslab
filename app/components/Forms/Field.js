// @flow
import React from 'react'
import * as R from 'ramda'
import type { NestedArray } from 'ramda'
import styled from 'styled-components'
import { Field as FinalFormField } from 'react-final-form'
import { composeValidators, required } from './validators'
import { fieldWrapperStyles, labelStyles, helpTextStyles, inputStyles, validationErrorStyles } from './styles'

export const Wrapper = styled.div`
	${fieldWrapperStyles};
`

export const Label = styled.label`
	${labelStyles};
`

export const HelpText = styled.h4`
	${helpTextStyles};
`

const Input = styled.input`
	${inputStyles};
`

export const ValidationError = styled.h4`
	${validationErrorStyles};
`

type Props = {
	label: string,
	name: string,
	validate?: Function | Array<Function>,
	helpText?: string,
	type?: string,
	required?: boolean,
}

export const arrayifyFunctions = (...things: NestedArray<Function>): Array<Function> => R.flatten(things)

const Field = (props: Props) => {
	const validators = props.validate
		? props.required ? composeValidators(required, props.validate) : composeValidators(props.validate)
		: undefined
	return (
		<FinalFormField
			name={props.name}
			validate={validators}
			render={({ input, meta }) => (
				<Wrapper>
					<Label required={props.required}>{props.label}</Label>
					<Input {...input} type={props.type} active={meta.active} />
					<HelpText>{props.helpText}</HelpText>
					{meta.error && meta.touched ? <ValidationError>{meta.error}</ValidationError> : null}
					{meta.submitError && !meta.dirtySinceLastSubmit ? <ValidationError>{meta.submitError}</ValidationError> : null}
				</Wrapper>
			)}
		/>
	)
}

Field.defaultProps = {
	type: 'input',
	required: false,
	helpText: '',
	validate: () => undefined,
}

export default Field
