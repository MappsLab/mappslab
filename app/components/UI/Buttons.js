// @flow
import React from 'react'
import styled from 'styled-components'

import { buttonStyles } from './styles'

export const ButtonWrapper = styled.button`
	${buttonStyles};
`

type ButtonProps = {
	type?: 'button' | 'submit' | 'reset',
}

export const Button = (props: ButtonProps) => <ButtonWrapper {...props} />

Button.defaultProps = {
	type: 'button',
}

type SubmitButtonProps = {
	disabled: boolean,
}

export const SubmitButton = (props: SubmitButtonProps) => <ButtonWrapper {...props} type="submit" />
