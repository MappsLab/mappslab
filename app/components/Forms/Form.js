// @flow
import React from 'react'
import type { Node } from 'react'
import { Form as FinalForm } from 'react-final-form'
import styled from 'styled-components'
import { SubmitButton } from '../UI/Buttons'
import { formWrapperStyles, formMessageStyles, formErrorStyles } from './styles'

const FormWrapper = styled.form`
	${formWrapperStyles};
`

export const FormMessage = styled.h4`
	${formMessageStyles};
`

export const FormError = styled.h4`
	${formErrorStyles};
`

/**
 * Form
 */

type Props = {
	onSubmit: (Object) => Promise<Object | void | Error>,
	message?: string | null,
	errorMessage?: string | null,
	submitButtonText?: string,
	showSubmitButton?: boolean,
	disabled?: boolean,
	validate?: (values: Object) => Object | Promise<Object>,
	children?: Node | Array<Node>,
	render?: (formProps: Object) => Node | Array<Node>,
}

const Form = (props: Props) => (
	<FinalForm
		{...props}
		render={(formProps) => {
			const { handleSubmit, submitError, pristine, submitting, submitErrors, submitFailed, dirtySinceLastSubmit } = formProps
			const error = props.errorMessage || submitError
			const { message } = props
			const submitFailedAndNotDirty = submitFailed && submitErrors && !dirtySinceLastSubmit
			const buttonDisabled = props.disabled || pristine || submitFailedAndNotDirty || submitting || false
			return (
				<FormWrapper disabled={props.disabled} onSubmit={handleSubmit}>
					{props.children || (props.render && props.render(formProps))}
					{message ? <FormMessage>{message}</FormMessage> : null}
					{error && !submitting && !dirtySinceLastSubmit ? <FormError>{error}</FormError> : null}
					{props.showSubmitButton && (
						<SubmitButton wide disabled={buttonDisabled}>
							{props.submitButtonText}
						</SubmitButton>
					)}
				</FormWrapper>
			)
		}}
	/>
)

Form.defaultProps = {
	message: undefined,
	errorMessage: undefined,
	submitButtonText: 'Submit',
	disabled: false,
	showSubmitButton: true,
	children: undefined,
	render: () => null,
	validate: () => ({}),
}

export default Form
