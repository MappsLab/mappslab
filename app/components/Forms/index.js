// @flow
import Field from './Field'
import Form from './Form'

module.exports = {
	Field,
	Form,
}
