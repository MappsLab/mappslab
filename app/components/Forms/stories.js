/* eslint-disable import/no-extraneous-dependencies */
import React, { Fragment } from 'react'
import { storiesOf } from '@storybook/react'
import { Field, Form } from 'Components/Forms'
import { Header2 } from 'Components/Text'
import { emailValidator } from 'Components/Forms/validators'
import { Annotation } from 'Storybook/components'
import { FORM_ERROR } from 'final-form'

const submitHandler = ({ email, password }, FormAPI) => {
	if (email === 'info@availableworks.net' && password === 'password') {
		alert('success')
		return undefined
	}
	return {
		[FORM_ERROR]: 'Email and password do not match.',
	}
}

storiesOf('Components/Forms', module).add('Basic Form (login)', () => (
	<Fragment>
		<Header2 align="center" weight="light">
			Nice to see you again.
		</Header2>
		<Form onSubmit={submitHandler}>
			<Field label="email" name="email" type="email" validate={emailValidator} />
			<Field label="password" name="password" type="password" />
		</Form>
		<Annotation>credentials are "info@availableworks.net" & "password"</Annotation>
	</Fragment>
))
