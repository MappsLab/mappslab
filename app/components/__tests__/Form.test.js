/* eslint-disable no-undef */
import React from 'react'
import { shallow, mount } from 'enzyme'
import { FORM_ERROR } from 'final-form'

import { Form, Field } from '../Forms/'
import { FormMessage, FormError } from '../Forms/Form'
import { SubmitButton, Button } from '../UI/Buttons'
import { ValidationError } from '../Forms/Field'
import { email, required } from '../Forms/validators'

describe('Form Component', () => {
	it('Mounts Correctly', () => {
		const wrapper = shallow(<Form />)
		expect(wrapper).toMatchSnapshot()
	})

	it('Supplies its own submit button', () => {
		const wrapper = mount(
			<Form onSubmit={() => {}}>
				<Field label="Email" name="email" type="email" validate={email} />
				<Field label="Password" name="password" type="password" />
			</Form>,
		)
		expect(wrapper.find(SubmitButton).length).toBe(1)
	})

	// TODO: This isn't working, i can't get the mock to be called..
	// it('Will not submit when regular buttons are pressed', () => {
	// 	const mockSubmitFn = jest.fn()
	// 	const wrapper = mount(
	// 		<Form
	// 			onSubmit={() => {
	// 				console.log('!!!')
	// 				mockSubmitFn()
	// 			}}
	// 		>
	// 			<Field label="Email" name="email" type="email" validate={email} />

	// 			<Button data-test="not-submit">Some other button</Button>
	// 		</Form>,
	// 	)

	// 	wrapper.find('input[name="email"]').simulate('change', { target: { value: 'joseph@joseph.com' } })
	// 	// console.log(wrapper.debug())
	// 	wrapper.find('button[type="submit"]').simulate('click')
	// 	expect(mockSubmitFn).toHaveBeenCalled()
	// })

	describe('Submission', () => {
		const mockSubmitFn = jest.fn()
		const wrapper = mount(
			<Form onSubmit={mockSubmitFn}>
				<Field label="Email" name="email" type="email" validate={email} />
				<Field label="Password" name="password" type="password" />
			</Form>,
		)

		beforeEach(() => {
			mockSubmitFn.calls = []
			wrapper.unmount()
			wrapper.mount()
		})

		it('Will not submit when pristine', () => {
			wrapper.find('button[type="submit"]').simulate('click')
			expect(mockSubmitFn).not.toHaveBeenCalled()
		})

		it('Will enable button if not pristine', () => {
			wrapper.find('input[name="email"]').simulate('change', { target: { value: 'joseph@joseph.com' } })
			expect(wrapper.find('button[type="submit"]').props().disabled).toEqual(false)

			// TODO: This test isn't working, i can't get the mock to be called...
			// wrapper.find('button[type="submit"]').simulate('click')
			// expect(mockSubmitFn).toHaveBeenCalled()
		})

		it('Will disable the button with a `disabled` prop', () => {
			wrapper.setProps({ disabled: true })
			wrapper.find('input[name="email"]').simulate('change', { target: { value: 'joseph' } })
			expect(wrapper.find('button[type="submit"]').props().disabled).toEqual(true)
		})

		it('Will disable the button if the form is invalid', () => {
			wrapper.find('input[name="email"]').simulate('change', { target: { value: 'joseph' } })
			expect(wrapper.find('button[type="submit"]').props().disabled).toEqual(true)
		})
	})

	describe('Form Messages & Validation Errors', () => {
		const generalMessage = 'A general form message'

		const errorMessage = 'There was an error'
		const returnSubmissionError = () => ({
			[FORM_ERROR]: errorMessage,
		})
		const wrapper = mount(
			<Form onSubmit={returnSubmissionError} message={generalMessage}>
				<Field name="firstName" label="name" />
			</Form>,
		)

		it('Renders general messages', () => {
			expect(wrapper.find(FormMessage).text()).toEqual(generalMessage)
		})

		it('Renders form submission errors', () => {
			wrapper.find('input[name="firstName"]').simulate('change', { target: { value: 'joseph' } })
			wrapper.find('form').simulate('submit')
			expect(wrapper.find(FormError).text()).toEqual(errorMessage)
		})
	})

	describe('Field validation & submission errors', () => {
		const emailMessage = 'Email submission error'
		const usernameMessage = 'Username submission error'
		const returnFieldErrors = () => ({
			email: emailMessage,
			username: usernameMessage,
		})

		const wrapper = mount(
			<Form onSubmit={returnFieldErrors}>
				<Field name="email" label="email" type="email" validate={email} />
				<Field name="username" label="username" validate={required} />
			</Form>,
		)

		const emailInput = wrapper.find('input[name="email"]')
		const usernameInput = wrapper.find('input[name="username"]')
		const form = wrapper.find('form')

		emailInput.simulate('change', { target: { value: 'wrong' } })
		emailInput.simulate('blur')
		expect(
			wrapper
				.find({ name: 'email' })
				.find(ValidationError)
				.text(),
		).toEqual('Please enter a valid email address')

		emailInput.simulate('change', { target: { value: 'valid@email.com' } })
		emailInput.simulate('blur')

		expect(wrapper.find({ name: 'email' }).find(ValidationError).length).toEqual(0)

		form.simulate('submit')

		expect(
			wrapper
				.find({ name: 'username' })
				.find(ValidationError)
				.text(),
		).toEqual('Required')
		usernameInput.simulate('change', { target: { value: 'myUsername' } })

		expect(wrapper.find({ name: 'username' }).find(ValidationError).length).toEqual(0)

		form.simulate('submit')

		expect(wrapper.text()).toContain(emailMessage)
		expect(wrapper.text()).toContain(usernameMessage)
	})

	describe('Form allows for re-submission after errors', () => {
		// console.log('Running test..')
		const invalidEmail = 'Invalid Email'
		const returnSubmitErrors = (values) => {
			if (values.email !== 'joseph@good-idea.studio') return { [FORM_ERROR]: invalidEmail }
			return undefined
		}
		const wrapper = mount(
			<Form onSubmit={returnSubmitErrors}>
				<Field name="email" label="email" type="email" validate={email} />
			</Form>,
		)

		const emailInput = wrapper.find('input[name="email"]')
		const form = wrapper.find('form')
		emailInput.simulate('change', { target: { value: 'josephkthomas@gmail.com' } })

		form.simulate('submit')
		expect(wrapper.find(FormError).text()).toBe(invalidEmail)
		expect(wrapper.find(SubmitButton).props().disabled).toBe(true)
		emailInput.simulate('change', { target: { value: 'joseph@good-idea.studio' } })
		expect(wrapper.find(SubmitButton).props().disabled).toBe(false)
	})
})
