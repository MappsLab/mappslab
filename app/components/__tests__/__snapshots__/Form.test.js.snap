// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form Component Mounts Correctly 1`] = `
.c1 {
  display: block;
  -webkit-transition: 0.2s;
  transition: 0.2s;
  border-radius: 2px;
  text-transform: uppercase;
  box-shadow: 0px 2px 3px 0px rgba(50,50,50,0.5);
  -webkit-letter-spacing: 0.06em;
  -moz-letter-spacing: 0.06em;
  -ms-letter-spacing: 0.06em;
  letter-spacing: 0.06em;
  text-transform: uppercase;
  font-size: 14px;
  margin: 14px auto;
  padding: 7px 14px;
  background-color: #6ac56d;
  color: black;
}

.c1:hover {
  background-color: #b2fab4;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  text-align: center;
}

<div>
  <form
    class="c0"
  >
    <button
      class="c1"
      disabled=""
      type="submit"
    >
      Submit
    </button>
  </form>
</div>
`;
